# Term Project - Statistics for Data Science
## Exploratory Data Analysis
### Group 6
* Kalyan Bondada
* Shitiz Kumar Goyal
* Carlos Hinrichsen Picand
* Priyanka Kamboj
* Katya Ramos


# Summary

This notebook has the following structure:
1. __General Information of Data Set__
2. __Exploratory Analysis__
3. __Classification Analysis__

# I. General Information of Data Set

### Default of Credit Card Clients Data Set

In our term project we will use the data set of default credit card clients, provided in the webpage of the UCI Machine Learning Repository: https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients

### Data Set Information:

1. The default variable, `default payment next month`, is the response variable (Y)

On the other hand, we have the following potential predictors:

1. `LIMIT_BAL`: Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit
2. `SEX`: (1 = male; 2 = female)
3. `EDUCATION`: (1 = graduate school; 2 = university; 3 = high school; 4 = others) 
4. `MARRIAGE`: (1 = married; 2 = single; 3 = others) 
5. `AGE`: (year). 
6. History of past payment. We tracked the past monthly payment records (from April to September, 2005) as follows: `PAY_0` = the repayment status in September, 2005; `PAY_2` = the repayment status in August, 2005; . . .;`PAY_6` = the repayment status in April, 2005. The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above. 
7. Amount of bill statement (NT dollar). `BILL_AMT1` = amount of bill statement in September, 2005; `BILL_AMT2` = amount of bill statement in August, 2005; . . .; `BILL_AMT6` = amount of bill statement in April, 2005. 
8. Amount of previous payment (NT dollar). `PAY_AMT1` = amount paid in September, 2005; `PAY_AMT2` = amount paid in August, 2005; . . .;`PAY_AMT6` = amount paid in April, 2005. 

__Below you will find a series of steps in order to understand the data:__

# II. Exploratory Analysis

## 1. Import Required Libraries

In [3]:
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import math
import random
import pandas as pd
plt.style.use('ggplot')

## 2. Upload the `default of credit card clients.xls` data to explore

In [4]:
import pandas as pd
odata = pd.read_excel('../data/raw/default of credit card clients.xls', header=[1],sheet_name="Data")

## 3. Exploring the raw data set

The names of the different variables is shown below.

In [5]:
# Extract column names
odata.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

Now, we will review the general characteristics of the dataset, and then review the data.

In [6]:
# Extract information of dataset
odata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
ID                            30000 non-null int64
LIMIT_BAL                     30000 non-null int64
SEX                           30000 non-null int64
EDUCATION                     30000 non-null int64
MARRIAGE                      30000 non-null int64
AGE                           30000 non-null int64
PAY_0                         30000 non-null int64
PAY_2                         30000 non-null int64
PAY_3                         30000 non-null int64
PAY_4                         30000 non-null int64
PAY_5                         30000 non-null int64
PAY_6                         30000 non-null int64
BILL_AMT1                     30000 non-null int64
BILL_AMT2                     30000 non-null int64
BILL_AMT3                     30000 non-null int64
BILL_AMT4                     30000 non-null int64
BILL_AMT5                     30000 non-null int64
BILL_AMT6               

Looking for the general description of the data

In [7]:
# Extract description
odata.describe()

ID       LIMIT_BAL           SEX     EDUCATION      MARRIAGE  \
count  30000.000000    30000.000000  30000.000000  30000.000000  30000.000000   
mean   15000.500000   167484.322667      1.603733      1.853133      1.551867   
std     8660.398374   129747.661567      0.489129      0.790349      0.521970   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22500.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                AGE         PAY_0         PAY_2         PAY_3         PAY_4  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean      35.485500     -0.016700     -0.133767     -0.166200     -0.220667   
std        9.217904      1.123802      1.197186      1.196868      1.169139   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

                  ...                  BILL_AMT4      BILL_AMT5  \
count             ...               30000.000000   30000.000000   
mean              ...               43262.948967   40311.400967   
std               ...               64332.856134   60797.155770   
min               ...             -170000.000000  -81334.000000   
25%               ...                2326.750000    1763.000000   
50%               ...               19052.000000   18104.500000   
75%               ...               54506.000000   50190.500000   
max               ...              891586.000000  927171.000000   

           BILL_AMT6       PAY_AMT1      PAY_AMT2      PAY_AMT3  \
count   30000.000000   30000.000000  3.000000e+04   30000.00000   
mean    38871.760400    5663.580500  5.921163e+03    5225.68150   
std     59554.107537   16563.280354  2.304087e+04   17606.96147   
min   -339603.000000       0.000000  0.000000e+00       0.00000   
25%      1256.000000    1000.000000  8.330000e+02     390.00000   
50%     17071.000000    2100.000000  2.009000e+03    1800.00000   
75%     49198.250000    5006.000000  5.000000e+03    4505.00000   
max    961664.000000  873552.000000  1.684259e+06  896040.00000   

            PAY_AMT4       PAY_AMT5       PAY_AMT6  default payment next month  
count   30000.000000   30000.000000   30000.000000                30000.000000  
mean     4826.076867    4799.387633    5215.502567                    0.221200  
std     15666.159744   15278.305679   17777.465775                    0.415062  
min         0.000000       0.000000       0.000000                    0.000000  
25%       296.000000     252.500000     117.750000                    0.000000  
50%      1500.000000    1500.000000    1500.000000                    0.000000  
75%      4013.250000    4031.500000    4000.000000                    0.000000  
max    621000.000000  426529.000000  528666.000000                    1.000000  

[8 rows x 25 columns]

Important things to notice:

__1. There are no NAN observations in the dataset__ 

__2. All the variables are saved as int. Clearly most of the variables are categorical (not considering cases like `LIMIT_BAL` and bill and payment amounts)__

__3. The ID variable is not relevant. It should be removed__

__4. We will change the name of the variable `default payment next month` to `DEFAULT`__

__5. In the variables `EDUCATION`, `MARRIAGE`, and the different categories of Payments, have more categories as the described in the dataset description webpage. We will need to review each variable in order to understand what is happening__

__The most important statistic that we can see is that only__ __$22.12$%__ __of the people are in the default category.__

In [8]:
# Showing the data
odata

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
0          1      20000    2          2         1   24      2      2     -1   
1          2     120000    2          2         2   26     -1      2      0   
2          3      90000    2          2         2   34      0      0      0   
3          4      50000    2          2         1   37      0      0      0   
4          5      50000    1          2         1   57     -1      0     -1   
5          6      50000    1          1         2   37      0      0      0   
6          7     500000    1          1         2   29      0      0      0   
7          8     100000    2          2         2   23      0     -1     -1   
8          9     140000    2          3         1   28      0      0      2   
9         10      20000    1          3         2   35     -2     -2     -2   
10        11     200000    2          3         2   34      0      0      2   
11        12     260000    2          1         2   51     -1     -1     -1   
12        13     630000    2          2         2   41     -1      0     -1   
13        14      70000    1          2         2   30      1      2      2   
14        15     250000    1          1         2   29      0      0      0   
15        16      50000    2          3         3   23      1      2      0   
16        17      20000    1          1         2   24      0      0      2   
17        18     320000    1          1         1   49      0      0      0   
18        19     360000    2          1         1   49      1     -2     -2   
19        20     180000    2          1         2   29      1     -2     -2   
20        21     130000    2          3         2   39      0      0      0   
21        22     120000    2          2         1   39     -1     -1     -1   
22        23      70000    2          2         2   26      2      0      0   
23        24     450000    2          1         1   40     -2     -2     -2   
24        25      90000    1          1         2   23      0      0      0   
25        26      50000    1          3         2   23      0      0      0   
26        27      60000    1          1         2   27      1     -2     -1   
27        28      50000    2          3         2   30      0      0      0   
28        29      50000    2          3         1   47     -1     -1     -1   
29        30      50000    1          1         2   26      0      0      0   
...      ...        ...  ...        ...       ...  ...    ...    ...    ...   
29970  29971     360000    1          1         1   34     -1     -1     -1   
29971  29972      80000    1          3         1   36      0      0      0   
29972  29973     190000    1          1         1   37      0      0      0   
29973  29974     230000    1          2         1   35      1     -2     -2   
29974  29975      50000    1          2         1   37      1      2      2   
29975  29976     220000    1          2         1   41      0      0     -1   
29976  29977      40000    1          2         2   47      2      2      3   
29977  29978     420000    1          1         2   34      0      0      0   
29978  29979     310000    1          2         1   39      0      0      0   
29979  29980     180000    1          1         1   32     -2     -2     -2   
29980  29981      50000    1          3         2   42      0      0      0   
29981  29982      50000    1          2         1   44      1      2      2   
29982  29983      90000    1          2         1   36      0      0      0   
29983  29984      20000    1          2         1   44     -2     -2     -2   
29984  29985      30000    1          2         2   38     -1     -1     -2   
29985  29986     240000    1          1         2   30     -2     -2     -2   
29986  29987     360000    1          1         2   35     -1     -1     -2   
29987  29988     130000    1          1         2   34      0      0      0   
29988  29989     250000    1          1         1   34      

## 4. Data Cleaning

As explained above, we will first drop the irrelevant variables: ID, and rename the Default variable:

In [9]:
# Dropping the columns
odata = odata.drop(['ID'], axis=1)

In [10]:
# Remaning the default variable name
odata.rename(columns={'default payment next month':'DEFAULT'},inplace=True)

In [11]:
odata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 24 columns):
LIMIT_BAL    30000 non-null int64
SEX          30000 non-null int64
EDUCATION    30000 non-null int64
MARRIAGE     30000 non-null int64
AGE          30000 non-null int64
PAY_0        30000 non-null int64
PAY_2        30000 non-null int64
PAY_3        30000 non-null int64
PAY_4        30000 non-null int64
PAY_5        30000 non-null int64
PAY_6        30000 non-null int64
BILL_AMT1    30000 non-null int64
BILL_AMT2    30000 non-null int64
BILL_AMT3    30000 non-null int64
BILL_AMT4    30000 non-null int64
BILL_AMT5    30000 non-null int64
BILL_AMT6    30000 non-null int64
PAY_AMT1     30000 non-null int64
PAY_AMT2     30000 non-null int64
PAY_AMT3     30000 non-null int64
PAY_AMT4     30000 non-null int64
PAY_AMT5     30000 non-null int64
PAY_AMT6     30000 non-null int64
DEFAULT      30000 non-null int64
dtypes: int64(24)
memory usage: 5.5 MB


Now we will review the variables that have more categories as explained in the data set information

First, `EDUCATION`:

In [12]:
odata.EDUCATION.value_counts()/odata.EDUCATION.size*100

2    46.766667
1    35.283333
3    16.390000
5     0.933333
4     0.410000
6     0.170000
0     0.046667
Name: EDUCATION, dtype: float64

We can see that the categories described in the dataset information, 1 to 4, corresponds to over $95\%$ of the data. Therefore we will remove the other categories 

In [13]:
odata = odata[(odata.EDUCATION!=0) & (odata.EDUCATION!=5) & (odata.EDUCATION!=6)]
odata.EDUCATION.value_counts()

2    14030
1    10585
3     4917
4      123
Name: EDUCATION, dtype: int64

Second, `MARRIAGE`:

In [14]:
odata.MARRIAGE.value_counts()/odata.MARRIAGE.size*100

2    53.299612
1    45.445962
3     1.072332
0     0.182094
Name: MARRIAGE, dtype: float64

We can see that the categories described in the dataset information, 1 to 3, corresponds to over $95\%$ of the data. Therefore we will remove the othe category 

In [15]:
odata = odata[(odata.MARRIAGE!=0)]
odata.MARRIAGE.value_counts()

2    15806
1    13477
3      318
Name: MARRIAGE, dtype: int64

Third, payment categories:
1. `PAY_0`

In [16]:
odata.PAY_0.value_counts()/odata.PAY_0.size*100

 0    48.981453
-1    19.029763
 1    12.371204
-2     9.148340
 2     8.918618
 3     1.081045
 4     0.256748
 5     0.081078
 8     0.064187
 6     0.037161
 7     0.030404
Name: PAY_0, dtype: float64

We can see that the categories __NOT__ described in the dataset information, -2 and 0, corresponds to over $55\%$ of the data. Therefore we can't remove these categories. We will need further analysis. 

Before this, we will review the other payments categories to see if this behaviour is the same in the other variables

2. `PAY_2`

In [17]:
odata.PAY_2.value_counts()/odata.PAY_2.size*100

 0    52.282017
-1    20.235803
 2    13.188744
-2    12.573900
 3     1.101314
 4     0.327692
 1     0.094591
 5     0.084457
 7     0.067565
 6     0.040539
 8     0.003378
Name: PAY_2, dtype: float64

3. `PAY_3`

In [18]:
odata.PAY_3.value_counts()/odata.PAY_3.size*100

 0    52.423905
-1    19.806763
-2    13.604270
 2    12.844161
 3     0.800649
 4     0.256748
 7     0.091213
 6     0.077700
 5     0.070944
 1     0.013513
 8     0.010135
Name: PAY_3, dtype: float64

4. `PAY_4`

In [19]:
odata.PAY_4.value_counts()/odata.PAY_4.size*100

 0    54.741394
-1    18.975710
-2    14.482619
 2    10.614506
 3     0.608088
 4     0.233100
 7     0.195939
 5     0.118239
 6     0.016891
 8     0.006757
 1     0.006757
Name: PAY_4, dtype: float64

5. `PAY_5`

In [20]:
odata.PAY_5.value_counts()/odata.PAY_5.size*100

 0    56.362961
-1    18.512888
-2    15.131246
 2     8.840918
 3     0.597953
 4     0.283774
 7     0.195939
 5     0.057430
 6     0.013513
 8     0.003378
Name: PAY_5, dtype: float64

6. `PAY_6`

In [21]:
odata.PAY_6.value_counts()/odata.PAY_6.size*100

 0    54.231276
-1    19.168271
-2    16.235938
 2     9.310496
 3     0.618222
 4     0.165535
 7     0.155400
 6     0.064187
 5     0.043917
 8     0.006757
Name: PAY_6, dtype: float64

We can see that the effect of these two categories is getting bigger in the early past payments (for example, `PAY_6`is from April 2005, and `PAY_0`is from September 2005) 

### 4.1 Payments Review

In order to understand the categories described above, we will review a subset of the data, particullarly the data related to statements and payments.

Let's subset the data:

In [22]:
odata_s = odata.copy()
# Subset the data
odata_s = odata_s.iloc[:,5:23]
odata_s.head(15)

PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  BILL_AMT2  BILL_AMT3  \
0       2      2     -1     -1     -2     -2       3913       3102        689   
1      -1      2      0      0      0      2       2682       1725       2682   
2       0      0      0      0      0      0      29239      14027      13559   
3       0      0      0      0      0      0      46990      48233      49291   
4      -1      0     -1      0      0      0       8617       5670      35835   
5       0      0      0      0      0      0      64400      57069      57608   
6       0      0      0      0      0      0     367965     412023     445007   
7       0     -1     -1      0      0     -1      11876        380        601   
8       0      0      2      0      0      0      11285      14096      12108   
9      -2     -2     -2     -2     -1     -1          0          0          0   
10      0      0      2      0      0     -1      11073       9787       5535   
11     -1     -1     -1     -1     -1      2      12261      21670       9966   
12     -1      0     -1     -1     -1     -1      12137       6500       6500   
13      1      2      2      0      0      2      65802      67369      65701   
14      0      0      0      0      0      0      70887      67060      63561   

    BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  \
0           0          0          0         0       689         0         0   
1        3272       3455       3261         0      1000      1000      1000   
2       14331      14948      15549      1518      1500      1000      1000   
3       28314      28959      29547      2000      2019      1200      1100   
4       20940      19146      19131      2000     36681     10000      9000   
5       19394      19619      20024      2500      1815       657      1000   
6      542653     483003     473944     55000     40000     38000     20239   
7         221       -159        567       380       601         0       581   
8       12211      11793       3719      3329         0       432      1000   
9           0      13007      13912         0         0         0     13007   
10       2513       1828       3731      2306        12        50       300   
11       8517      22287      13668     21818      9966      8583     22301   
12       6500       6500       2870      1000      6500      6500      6500   
13      66782      36137      36894      3200         0      3000      3000   
14      59696      56875      55512      3000      3000      3000      3000   

    PAY_AMT5  PAY_AMT6  
0          0         0  
1          0      2000  
2       1000      5000  
3       1069      1000  
4        689       679  
5       1000       800  
6      13750     13770  
7       1687      1542  
8       1000      1000  
9       1122         0  
10      3738        66  
11         0      3640  
12      2870         0  
13      1500         0  
14      3000      3000

The first thing is to understand the logic in the payment information: As credit card users, we know that regullarly we receive our statement in one month, let's say $May 15^{th}$, and the due dat is usually one month before, in this case $June 15^{th}$. So usually there is a lag in time about our payment information, and our dataset should reflect that. For this purpose, let's look the individual #11 of the table above:  

* The variable `BILL_AMT2` is the bill amount of August 2005, which is $21670$. Following the logic, we should pay this bill statement on September 2005, which is the variable `PAY_AMT1` and corresponds to $21818$. In this case the bill was full payed, so the variable `PAY_2` should reflect that. In fact is true because the uput of the variable is -1, therefore the bill way pay in time. If the see the other statements and payments, the logic remains.

Let's see this effect in a plot:


In [23]:
plt.rc('figure', figsize=(8, 8))
pl_pay0 = sns.scatterplot(x="BILL_AMT2", y="PAY_AMT1",hue="PAY_2", size="PAY_2",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT2, odata_s.BILL_AMT2, '--r')

AttributeError: module 'seaborn' has no attribute 'scatterplot'

We can see that most of the cases are plotted correctly according to tha variable `PAY_1`. Additionally, there are cases where the `BILL_AMT2` is negative, therefore it's a credit or it's an error. This same observations is shown in the following plots. 

Now let's be suspicuious about this, and think that the variables are defined one by one, I mean that the variables `BILL_AMT1`is related to `PAY_AMT1`and `PAY_0`. The plot is the following:



In [ ]:
plt.rc('figure', figsize=(8, 8))
pl_pay0 = sns.scatterplot(x="BILL_AMT1", y="PAY_AMT1",hue="PAY_0", size="PAY_0",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT1, odata_s.BILL_AMT1, '--r')

Nos we can see that the effect is worst here, because there are more observations labelled incorrectly. So, in that case we will assume that the correct logic is the one of the first plot.

Let's plot the rest of them:

Related to other `PAY` variables:

In [ ]:
plt.rc('figure', figsize=(15, 15))
plt.subplot(2,2,1)
pl_pay2 = sns.scatterplot(x="BILL_AMT3", y="PAY_AMT2",hue="PAY_3", size="PAY_3",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT3, odata_s.BILL_AMT3, '--r')
plt.subplot(2,2,2)
pl_pay3 = sns.scatterplot(x="BILL_AMT4", y="PAY_AMT3",hue="PAY_4", size="PAY_4",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT4, odata_s.BILL_AMT4, '--r')
plt.subplot(2,2,3)
pl_pay4 = sns.scatterplot(x="BILL_AMT5", y="PAY_AMT4",hue="PAY_5", size="PAY_5",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT5, odata_s.BILL_AMT5, '--r')
plt.subplot(2,2,4)
pl_pay5 = sns.scatterplot(x="BILL_AMT6", y="PAY_AMT5",hue="PAY_6", size="PAY_6",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT6, odata_s.BILL_AMT6, '--r')

As the first plot, we can see the same structure here. 

Due the lag, we can't use the variable `PAY_AMT6`.

Considering the previous analysis, we can say:

* Considering that the information of the variables `PAY` are a transformation of information of other variables, we should eliminate the variables `PAY_AMT` and `BILL_AMT`. 
* Because of the lag, it's not possible to reproduce all `PAY` variables, but these have more relevant information than `PAY_AMT` and `BILL_AMT` variables. 

Before doing that, let's plot the rest of the plots defined by the variables without the lag, just to be sure about the effects.

In [ ]:
plt.rc('figure', figsize=(15, 15))
plt.subplot(2,3,1)
sns.scatterplot(x="BILL_AMT2", y="PAY_AMT2",hue="PAY_2", size="PAY_2",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT2, odata_s.BILL_AMT2, '--r')
plt.subplot(2,3,2)
sns.scatterplot(x="BILL_AMT3", y="PAY_AMT3",hue="PAY_3", size="PAY_3",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT3, odata_s.BILL_AMT3, '--r')
plt.subplot(2,3,3)
sns.scatterplot(x="BILL_AMT4", y="PAY_AMT4",hue="PAY_4", size="PAY_4",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT4, odata_s.BILL_AMT4, '--r')
plt.subplot(2,3,4)
sns.scatterplot(x="BILL_AMT5", y="PAY_AMT5",hue="PAY_5", size="PAY_5",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT5, odata_s.BILL_AMT5, '--r')
plt.subplot(2,3,5)
sns.scatterplot(x="BILL_AMT6", y="PAY_AMT6",hue="PAY_6", size="PAY_6",sizes=(20, 300), palette="Set2",legend="full", alpha=0.5,data=odata_s)
plt.plot(odata_s.BILL_AMT6, odata_s.BILL_AMT6, '--r')

We have the same effect, therefore...

In [ ]:
# Dropping the variables
odata = odata.drop(['BILL_AMT1','BILL_AMT2','BILL_AMT3','BILL_AMT4','BILL_AMT5','BILL_AMT6','PAY_AMT1','PAY_AMT2','PAY_AMT3','PAY_AMT4','PAY_AMT5','PAY_AMT6'], axis=1)

In [ ]:
odata.head()

## 5. Data Transformation

We will do a few transformations. Mainly to create categorical variables and create another two variables which is the sum of the different `PAY` variables. The logic is the lower this number is, the better chances is not no be on default. 

Before doing any transformation, let's notice the meaning of the different `PAY` variables:

* __-2:__ No Consumption
* __-1:__ Pay duly
* __0:__ The use of revolving credit
* __N:__ Payment delay for N months (N=1,2,...,8)
* __9:__ Payment delay for 9 months and above

Therefore, more than categorical, `PAY` variables is like a scale. Basically because these numbers could reflect the risk that the bank could take. A client with no consumption means that the bank has no risk, a client with consumption buy paying duly has risk, but very low, and the same logic with the rest of the number above. Consequently, the lower the number, better for the bank considering risk.


#### 5.1. Change "Sex", "Education", "Marriage", and "Default" variables into categorical type. 

In order to check that the transformation is working, we will cretae other variables. Once we check that the trasnformation is right, we will drop the original variables 

In [ ]:
odata['SEX_C'] = odata['SEX'].apply(lambda x: 'Male' if x == 1 else 'Female')
odata['SEX_C'] = pd.Categorical(odata.SEX_C,categories=["Male","Female"])
odata['EDUCATION_C'] = odata['EDUCATION'].apply(lambda x: 'Graduate' if x == 1 else ('University' if x==2 else ('High School' if x==3 else "Others")))
odata['EDUCATION_C'] = pd.Categorical(odata.EDUCATION_C,categories=["Graduate","University","High School","Others"])
odata['MARRIAGE_C'] = odata['MARRIAGE'].apply(lambda x: 'Married' if x == 1 else ('Single' if x==2 else "Others"))
odata['MARRIAGE_C'] = pd.Categorical(odata.MARRIAGE_C,categories=["Married","Single","Others"])
odata['DEFAULT_C'] = odata['DEFAULT'].apply(lambda x: 'Default' if x == 1 else 'No Default')
odata['DEFAULT_C'] = pd.Categorical(odata.DEFAULT_C,categories=["Default","No Default"])
odata.head()

The transformation is working properly, therefore we will drop the original variables 

In [ ]:
# Dropping the variables
odata = odata.drop(['SEX','EDUCATION','MARRIAGE','DEFAULT'], axis=1)
odata.head()

The last transformation is to create categories for the Age according to statistics Canada: https://www.statcan.gc.ca/eng/concepts/definitions/age2

The idea is no be able to analize better the relation between the variables. The categories are:

1. __Youth:__ 15-24 years
2. __Adults:__ 25-64 years
3. __Senior:__ Over 65 years

In [ ]:
# Creating the categories
age_cat = pd.cut(odata.AGE, [14, 24, 64,100], include_lowest=False)
odata['AGE_C'] = age_cat
odata['AGE_C'] = pd.Categorical(odata.AGE_C)
odata.head()

Now let's drop the original variables

In [ ]:
# Dropping the variable
odata = odata.drop(['AGE'], axis=1)
odata.head()

#### 5.2. Create a new variables which is the sum of the `PAY` variables

The logic is the lower this number is, the better chances is not no be on default.

In [ ]:
# Creating new variable
odata['PAY_SUM'] = odata.PAY_0 + odata.PAY_2 + odata.PAY_3 + odata.PAY_4 + odata.PAY_5 + odata.PAY_6
odata.head()

Now we are ready to analyze the data

## 6. Data Analysis

#### 6.1. Relation between `DEFAULT_C`and `LIMIT_BAL` and `PAY_SUM`. 


First, let's count again the proportion of defaults:

In [ ]:
odata.groupby(['DEFAULT_C']).size()/len(odata.DEFAULT_C)*100

After the data cleaning, the proportion is practically the same.

Nos the plotting:

In [ ]:
plt.rc('figure', figsize=(15, 8))
plt.subplot(1,2,1)
sns.boxplot(x="DEFAULT_C", y="LIMIT_BAL",data=odata)
plt.subplot(1,2,2)
sns.boxplot(x="DEFAULT_C", y="PAY_SUM",data=odata)

Looking the plots above, it's easy to see that the new variable, `PAY_SUM` could explain better the default than the `LIMIT_BAL`

Now let's plot all the `PAY` variables:

In [ ]:
plt.rc('figure', figsize=(15, 15))
plt.subplot(2,3,1)
sns.boxplot(x="DEFAULT_C", y="PAY_0",data=odata)
plt.subplot(2,3,2)
sns.boxplot(x="DEFAULT_C", y="PAY_2",data=odata)
plt.subplot(2,3,3)
sns.boxplot(x="DEFAULT_C", y="PAY_3",data=odata)
plt.subplot(2,3,4)
sns.boxplot(x="DEFAULT_C", y="PAY_4",data=odata)
plt.subplot(2,3,5)
sns.boxplot(x="DEFAULT_C", y="PAY_5",data=odata)
plt.subplot(2,3,6)
sns.boxplot(x="DEFAULT_C", y="PAY_6",data=odata)

It's difficult to see a clear trend. But we can somewhat infer:

* In general the "No Defaults" by month are with number below 0, which is correct
* There are months when the number are represented as pay in time, but they are labelled as default. That's because in one month you can't see the whole behaviour on an individual
* The overall, `PAY_SUM`, could be enough to represent the default variable, but if needed we can use every individual element of `PAY`

#### 6.2. Relation between `DEFAULT_C` and `AGE_C`.

In [ ]:
odata.groupby(['AGE_C','DEFAULT_C']).size().unstack()/len(odata.DEFAULT_C)*100

Most of the cases occur when the persons are adults, $\approx 90$%, and the proportion of defaults it's almost the same as the whole dataset, therefore this age group could represent the dataset.

#### 6.3. Relation between `DEFAULT_C` and `SEX_C`.

In [ ]:
odata.groupby(['SEX_C','DEFAULT_C']).size().unstack()/len(odata.DEFAULT_C)*100

We can see that there are more females than males ($60$% vs $40$%). Additionally, the males have a default tendency higher than females. Therefore, this variable could be good for the logistic regression. 

#### 6.4. Relation between `DEFAULT_C` and `EDUCATION_C`.

In [ ]:
odata.groupby(['EDUCATION_C','DEFAULT_C']).size().unstack()/len(odata.DEFAULT_C)*100

It's not uncommon to think that the highest proportion of Default comes with lower education. Therefore, this also could be a good explanatory variable

#### 6.5. Relation between `DEFAULT_C` and `MARRIAGE_C`.

In [ ]:
odata.groupby(['MARRIAGE_C','DEFAULT_C']).size().unstack()/len(odata.DEFAULT_C)*100

Highest proportion comes to single persons.

#### 6.6. Relation between `DEFAULT_C` with `SEX_C` and `AGE_C`.

In [ ]:
odata.groupby(['SEX_C','AGE_C','DEFAULT_C']).size().unstack()/len(odata.DEFAULT_C)*100

Here we can also see that the proportion of default of adult males is bigger than females. 

Additionally, youth females have higher proportion than youth males.

#### 6.7. Relation between `DEFAULT_C` with `EDUCATION_C` and `MARRIAGE_C`.

In [ ]:
odata.groupby(['EDUCATION_C','MARRIAGE_C','DEFAULT_C']).size().unstack()/len(odata.DEFAULT_C)*100

Most of the married couples hold a University degree, and the single persons in this group are almost the same quantity ($\approx 23$%). University singles have more tendency to default than graduate single ones.

In the married couples, the the University ones have more tendency to default than graduate ones.

#### 6.8. Relation between `LIMIT_BAL` and `PAY_SUM`.

In [ ]:
plt.rc('figure', figsize=(7, 7))
sns.scatterplot(x="PAY_SUM",y="LIMIT_BAL", data=odata)

It's interesting to see that highest `LIMIT_BAL` not always have lowest `PAY_SUM`, which means a huge risk for the bank (look the point on the far right with a `PAY_SUM` over 30 and `LIMIT_BAL` over 500000). 

We can also see that there are scenarios with lowest `LIMIT_BAL` and very low default scenarios (low `PAY_SUM`). In this scenarios there is a potential opportunity to offer to the clients to upgrade the `LIMIT_BAL` according to their good payments behavior.

We can see the relation between `LIMIT_BAL `with every `PAY` variable using a correlation matrix.

In [ ]:
# Copy dataset 
odata_p =odata.copy()
# Subset according
odata_p = odata_p.loc[:, ["LIMIT_BAL","PAY_0","PAY_2","PAY_3","PAY_4","PAY_5","PAY_6","PAY_SUM"]]
plt.figure(figsize=(12, 8))
sns.heatmap(odata_p.corr().round(4), annot=True,linewidths=1.0,vmin=-1,vmax=1,cmap="coolwarm")

It's interesting to see that each individual `PAY` category has a higher correlation with its adyacent neighbor. In other words, the highest correlation of `PAY_0` is with `PAY_1`, `PAY_2` is with `PAY_3`, etc. So, your payment behaviour has a trend between the months, if you pay on time one month it's likely that you will pay on time the next month. The same case when you are not paying on time.

Another thing that we have to notice is that `PAY_SUM` has a higher correlation to `LIMIT_BAL` than any other individual payment categories.

#### 6.9. Relation between `LIMIT_BAL` and `PAY_SUM` with `AGE_C`.

In [ ]:
plt.rc('figure', figsize=(15, 7))
plt.subplot(1,2,1)
sns.boxplot(x="AGE_C",y="LIMIT_BAL", data=odata)
plt.subplot(1,2,2)
sns.boxplot(x="AGE_C",y="PAY_SUM", data=odata)

From the boxplot we can see that with more age is more likely to have a higher `LIMIT_BAL`. This makes sense, because usually with age people become more responsible financially.

Additionally, younger people has more chances to default, which is related to `PAY_SUM`.

#### 6.10. Relation between `LIMIT_BAL` and `PAY_SUM` with `SEX_C`.

In [ ]:
plt.rc('figure', figsize=(15, 7))
plt.subplot(1,2,1)
sns.boxplot(x="SEX_C",y="LIMIT_BAL", data=odata)
plt.subplot(1,2,2)
sns.boxplot(x="SEX_C",y="PAY_SUM", data=odata)

In this case is not possible to distingish between male and female with the `LIMIT_BAL`.

As previous analysis, males have a higher tendency to default than females.

#### 6.11. Relation between `LIMIT_BAL` and `PAY_SUM` with `EDUCATION_C`.

In [ ]:
plt.rc('figure', figsize=(15, 7))
plt.subplot(1,2,1)
sns.boxplot(x="EDUCATION_C",y="LIMIT_BAL", data=odata)
plt.subplot(1,2,2)
sns.boxplot(x="EDUCATION_C",y="PAY_SUM", data=odata)

Higher education means higher `LIMIT_BAL`.

On the other side, University and High School people have a higher tendency to default.

# III. Classification Analysis

For this section, we will use the following dataset:

In [ ]:
# Data for the analysis
data_a = odata.copy()
data_a.info()

## 1. Logistic Regression

## 2. KNN Classifier